File: sdks/typescript/src/resources/purchase_orders.ts
import { AxiosInstance } from 'axios';
import { PaginatedResponse } from '../types';

export interface PurchaseOrder {
    id: string;
    supplier: string;
    items: PurchaseOrderItem[];
    status: string;
    total: number;
    created_at: string;
    updated_at: string;
}

export interface PurchaseOrderItem {
    product: string;
    quantity: number;
    cost_price: number;
}

export class PurchaseOrders {
    constructor(private client: AxiosInstance) { }

    async list(page: number = 1, pageSize: number = 20, status?: string): Promise<PaginatedResponse<PurchaseOrder>> {
        const params: any = { page, page_size: pageSize };
        if (status) params.status = status;
        const response = await this.client.get('/purchase-orders/', { params });
        return response.data;
    }

    async get(poId: string): Promise<PurchaseOrder> {
        const response = await this.client.get(`/purchase-orders/${poId}/`);
        return response.data;
    }

    async create(data: Partial<PurchaseOrder>): Promise<PurchaseOrder> {
        const response = await this.client.post('/purchase-orders/', data);
        return response.data;
    }

    async update(poId: string, data: Partial<PurchaseOrder>): Promise<PurchaseOrder> {
        const response = await this.client.patch(`/purchase-orders/${poId}/`, data);
        return response.data;
    }

    async receive(poId: string, items: PurchaseOrderItem[]): Promise<PurchaseOrder> {
        const response = await this.client.post(`/purchase-orders/${poId}/receive/`, { items });
        return response.data;
    }
}

File: sdks/rust/src/resources/purchase_orders.rs
use crate::models::{PaginatedResponse, PurchaseOrder};
use crate::Puxbay;
use crate::PuxbayError;
use reqwest::Method;
use serde::Serialize;

pub struct PurchaseOrdersResource<'a> {
    pub(crate) client: &'a Puxbay,
}

impl<'a> PurchaseOrdersResource<'a> {
    pub fn new(client: &'a Puxbay) -> Self {
        Self { client }
    }

    pub async fn list(&self, page: i32, status: Option<&str>) -> Result<PaginatedResponse<PurchaseOrder>, PuxbayError> {
        let mut query = format!("purchase-orders/?page={}", page);
        if let Some(s) = status {
            query.push_str(&format!("&status={}", s));
        }

        self.client.request(
            Method::GET,
            &query,
            None::<()>,
        ).await
    }

    pub async fn get(&self, po_id: &str) -> Result<PurchaseOrder, PuxbayError> {
        self.client.request(
            Method::GET,
            &format!("purchase-orders/{}/", po_id),
            None::<()>,
        ).await
    }

    pub async fn create(&self, po: &PurchaseOrder) -> Result<PurchaseOrder, PuxbayError> {
        self.client.request(
            Method::POST,
            "purchase-orders/",
            Some(po),
        ).await
    }

    pub async fn update(&self, po_id: &str, po: &PurchaseOrder) -> Result<PurchaseOrder, PuxbayError> {
        self.client.request(
            Method::PATCH,
            &format!("purchase-orders/{}/", po_id),
            Some(po),
        ).await
    }

    pub async fn receive(&self, po_id: &str, items: Vec<serde_json::Value>) -> Result<PurchaseOrder, PuxbayError> {
        #[derive(Serialize)]
        struct ReceiveRequest {
            items: Vec<serde_json::Value>,
        }

        self.client.request(
            Method::POST,
            &format!("purchase-orders/{}/receive/", po_id),
            Some(&ReceiveRequest { items }),
        ).await
    }
}
